First item tracker

sdilts opened the pull request:
ImageTracker's initial methods have been laid out.
+ Check the methods to make sure they follow the correct interface.
+ BlockingQueue is used for the queue
   - insertion and removal are blocking.

The Blob class has been inserted as a placeholder so that ImageTracker.java can compile

sdilts committed:
Added imageTracker and basic tests
+ Uses generics in order to compile: interface still needs to be finalized

sdilts committed:
Added shell of Blob cl

sdilts committed:
Added doc comment to blob class

+ Class doesn't do anything yet
+ Needed as placeholder for ImageTracker class

sdilts committed:
Adapted code to use BufferedImage and Blob objects. Updated javadoc.

Testing code:
+ Some parameters to constructors in the testing module are assumed
  reasonable defaults. May change
Normal code:
+ The blocking functions used by the queue throw Interrupt exceptions. These are
  handled inside the methods that call these functions. When an
  exception is thrown, the program will exit

sagersmith8 commented on src/main/java/csci432/Blob.java:
```diff
@@ -0,0 +1,7 @@
+ * Blob class for keeping track of moving items
+ **/
+public class Blob {
+}
```
github wants a newline

sagersmith8 replied:
I will be adding to this, but this will work for now

sdilts replied:
At the end of the file?

sagersmith8 replied:
Yea github wants all files to end in a newline

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+ *
+ **/
+public class ItemTracker<T> {
+    BlockingQueue<BufferedImage> queue;
```
Make this be `Queue<BufferedImage> queue` since `BlockingQueue` `extends` `Queue`

sdilts replied:
The blocking methods aren't included in the Queue interface. Or is it too early to think about it?

sagersmith8 replied:
Ah okay then you are correct

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+	/*
+	while(true) {
+	    if(!queue.isEmpty) {
+		T item = queue.take();
```
If you wanted this to work it would be: `BufferedImage item = queue.poll();`

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+     **/
+    public void run() {
+	/*
+	while(true) {
```
` while(!queue.isEmpty())`

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+    public void run() {
+	/*
+	while(true) {
+	    if(!queue.isEmpty) {
```
Remove this if

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+	    if(!queue.isEmpty) {
+		T item = queue.take();
+		processItem(item);
+	    }
```
Remove this bracket

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+	    //errors when interrupted:
+	} catch(InterruptedException e) {
+	    e.printStackTrace();
+	    System.exit(1);
```
Probably don't want the entire program to blow up if a thing can't be inserted, I would remove this line

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @param item a list whose contents are to be inserted into the processing queue
+     **/
+    public void insertItem(BufferedImage[] item) {
```
Would make this take `List<BufferedImage> items` instead. `Lists` are more stable than `Arrays` and can be made thread safe if we want to later

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+    public void processItem() {
+	try {
+	//BlockingQueue.take() is a blocking function
+	    processItem(queue.take());
```
`processItem(queue.poll());` I believe?

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+	    processItem(queue.take());
+	} catch(InterruptedException e) {
+	    e.printStackTrace();
+	    System.exit(1);
```
Again probably don't want to program to explode, just want it to notify us that something happened

sagersmith8 commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,97 @@
+    private void processItem(BufferedImage item) {
+	
+    }
+}
```
Github wants a newline

sagersmith8 commented on src/test/java/csci432/ItemTrackerTest.java:
```diff
@@ -0,0 +1,28 @@
+	ItemTracker t = new ItemTracker();
+	//TYPE_INT_ARGB is a placeholder 
+	t.insertItem(new BufferedImage(500,500, BufferedImage.TYPE_INT_ARGB));
+	assert(t.isProcessing() == true);
```
Very impressed on how close you got here but:

`assertThat(t.isProcessing(), is(true));`

sagersmith8 commented on src/test/java/csci432/ItemTrackerTest.java:
```diff
@@ -0,0 +1,28 @@
+	//TYPE_INT_ARGB is a placeholder 
+	t.insertItem(new BufferedImage(500,500, BufferedImage.TYPE_INT_ARGB));
+	t.processItem();
+	assert(t.isProcessing() == false);
```
Same as above

sagersmith8 commented on src/test/java/csci432/ItemTrackerTest.java:
```diff
@@ -0,0 +1,28 @@
+	t.processItem();
+	assert(t.isProcessing() == false);
+    }
+}
```
Github wants a newline

sagersmith8 commented on src/test/java/csci432/ItemTrackerTest.java:
```diff
@@ -0,0 +1,28 @@
+    public void insertItemTest() {
+	ItemTracker t = new ItemTracker();
+	//TYPE_INT_ARGB is a placeholder 
+	t.insertItem(new BufferedImage(500,500, BufferedImage.TYPE_INT_ARGB));
```
Probably want a way to verify that this method actually works. That may involve checking the size of the queue, or checking that the last item is the item that you just added. Either way we will want a unit test around that

sdilts committed:
Implemented changes as suggested on pull request page

+ newlines were added to the ends of files
ItemTracker.java:
+ Queue interface is now targeted instead of the BlockingQueue
  interface. Comments were left with the blocking equivalent
+ Run method has be changed and uncommented
+ Try/catch blocks have been removed: Queue interface doesn't
  throw exceptions
+ isProcessing() changed to itemsWaiting() to better reflect the original
  intent of the method, which is to show the status of the image queue
Testing Code:
+ Tests were amended

sdilts committed:
Implemented changes as suggested on pull request page

+ newlines were added to the ends of files
ItemTracker.java:
+ Queue interface is now targeted instead of the BlockingQueue
  interface. Comments were left with the blocking equivalent
+ Run method has be changed and uncommented
+ Try/catch blocks have been removed: Queue interface doesn't
  throw exceptions
+ isProcessing() changed to itemsWaiting() to better reflect the original
  intent of the method, which is to show the status of the image queue
Testing Code:
+ Tests were amended

sdilts committed:
Merge branch 'firstItemTracker' of https://github.com/sdilts/image_tracker into firstItemTracker

sdilts committed:
Updated javadoc comment to match

sagesmith-wf commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,84 @@
+	    BufferedImage item = queue.poll();
+	    processItem(item);
+	} 
+	
```
remove newline

sagesmith-wf commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * The method that does stuff
+     **/
+    public void run() {
```
Switch access modifier to `private`

sagesmith-wf commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,84 @@
+	queue = new LinkedBlockingQueue<BufferedImage>();
+	currentBlobs = new LinkedList<Blob>();
+    }
+
```
Add this method please:

    /**
    * Starts the ImageTracker service
    */
    public void start() {
         new Thread(() -> {
            run();
        }).start();
    }

sdilts replied:
Does the class need to implement` Runnable` or another interface?

sagesmith-wf replied:
Nah this is the jankety way to do it. It just makes a the runnable using a lambda because I am lazy

sagesmith-wf commented on src/main/java/csci432/ItemTracker.java:
```diff
@@ -0,0 +1,84 @@
+     * The method that does stuff
+     **/
+    public void run() {
+	while(!queue.isEmpty()) {
```
I think I lied here, super super sorry. I think what you originally had commented out was actually correct...

sdilts committed:
Added start() method and fixed run() method

+ Previous version of run() was correct

sagesmith-wf commented:
+1 Looks good, pending Travis and Codecov, will merge
